### DISCUSSION TITLE
How to write database adapter for specific database?

### DISCUSSION THREAD

**Discussion by ahmed-yasen-va** [ory/hydra#2269] [Q&A] [answered]
I want to make hydra support the [aerospike database](https://www.aerospike.com/) but I couldn't find any documentation for writing a database adapter.

**Comment by aeneasr** [comment, marked as answer]
We no longer support custom adapters due to the complexity involved. [Writing your own adapter is a bad idea](https://github.com/ory/hydra/issues/1627#issuecomment-549563206):

>This topic has come up time and time again, the original plugin interface was written to support a certain use case we no longer support.
>
>Writing your own database adapter is **a very bad idea and serious security issue** because:
>
>1. Our internal interfaces change often, at least with every schema change. Writing your own adapter means in 99.9% of the cases that you will stay with the version your adapter works for. This is, for obvious reasons, terrible.
>2. The SQL datastore supports advanced logic such as `ON CASCADE` statements that are necessary to implement certain security-related workflows such as properly removing a chain of tokens when one of them is being revoked or a client is deleted or consent is withdrawn. We obviously have some tests for that in place internally, but not all of them are exposed and they also change depending on internal schedules.
>
>Most "we want to roll our own adapter" boil down to statements such as: "well we use (insert k/v, document datastore here) and we don't want to operate another system". If that's the reason you want to implement a new datastore - don't. It's riskier, more work, and more longterm maintenance to keep everything up to date than it is to just spin up a Google Cloud SQL or Amazon RDS instance, and way, way more secure.
>
>ORY Hydra is a piece of software that **needs** a proper SQL datastore with consistency guarantees. Neither a k/v nor a document store can acomodate that. Unless you're trying to add something like MariaDB or some other SQL language (hopefully not Oracle) it's an absolute no-no anyways!
>
>In conclusion, we will not change this and will most likely deprecate plugins alltogether at some point.

**Reply by ahmed-yasen-va** [reply]
Thanks for you @aeneasr, but where I can find the adapters of postgres or mySql databases ?

**Reply by aeneasr** [reply]
In the persistence/sql directory

**Reply by isra270** [reply]
if any way make hydra store his data in Aerospike ?

**Reply by zepatrik** [reply]
Maybe @ahmed-yasen-va will create a fork that only has a aerospike persister, so then you could use that. But unless it is compatible with one of our supported SQL adapters there is no way.

**Reply by isra270** [reply]
> Maybe @ahmed-yasen-va will create a fork that only has a aerospike persister, so then you could use that. But unless it is compatible with one of our supported SQL adapters there is no way.

Thanks @zepatrik  we will wait aerospike persister

**Reply by aeneasr** [reply]
A fork is an option, but honestly, writing your own DBAL for ORY Hydra is an extremely bad idea. I have explained why in this issue comment: https://github.com/ory/hydra/issues/1627#issuecomment-549563206

**Reply by isra270** [reply]
thank you @aeneasr for clarify.

**Reply by ahmed-yasen-va** [reply]
Thanks for your answer, but I have to build a custom db adapter, so If there is a way please tell us how.
@aeneasr 
@zepatrik

**Reply by aeneasr** [reply]
No there is not and never will be.

**Reply by isra270** [reply]
@aeneasr can you recommend authorization server support Aerospike to use it instead of build own Adapter in hydra authorization server?

**Comment by Mafiree-dev** [comment]
We don’t support custom database adapters, and I think that’s the right tradeoff for a security-sensitive product. A database adapter sounds flexible, but it would add a lot of hidden complexity around token storage, consistency, revocation, and operational support.

If you want to use Aerospike, the safer path is to keep Hydra’s core storage on a supported SQL database and integrate Aerospike around it for the parts it fits best.

- **Keep core auth state in SQL**: that preserves the guarantees Hydra needs for clients, consent, sessions, and token lifecycle.
- **Use Aerospike for adjacent workloads**: caching, high-speed lookups, or event-driven data where strong relational semantics matter less.
- **Avoid custom adapter risk**: a new storage layer changes upgrade paths, support burden, and failure modes.
- **Consider upstream-friendly alternatives**: use the supported extension points instead of replacing the storage engine.

Think of it like building a house: you can customize the furniture, but you don’t want to swap out the foundation unless the whole structure was designed for it.

The shortest path is usually the safest one.
